### PR TITLE
leases: add more details on API Server Identity

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim <andrewsy@google.com>

Follow-up to https://github.com/kubernetes/website/pull/37921#discussion_r1036046029

### CODE REVIEW

File: content/en/docs/concepts/architecture/leases.md [hunk 1]
```diff
@@ -38,3 +38,20 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+```
+$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
+NAME                                        HOLDER                                                                           AGE
+kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
+kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
+kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
+```
+
+The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be
+configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname
```
Comment (enj) [inline content/en/docs/concepts/architecture/leases.md:54]: Can we document the use of the `kubernetes.io/hostname` label as well?

File: content/en/docs/concepts/architecture/leases.md [hunk 2]
```diff
@@ -38,3 +38,20 @@ rest of the system. While not particularly useful on its own, this provides a me
...
 Existence of kube-apiserver leases enables future capabilities that may require coordination between
 each kube-apiserver.
+
+You can inspect Leases owned by each kube-apiserver by checking for lease objects in the `kube-system` namespace
+with the name `kube-apiserver-<sha256-hash>`. Alternatively you can use the label selector `k8s.io/component=kube-apiserver`:
+
+```
+$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
+NAME                                        HOLDER                                                                           AGE
+kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
```
Comment (enj) [inline content/en/docs/concepts/architecture/leases.md:48]: Should we include the YAML of one of these as an example?  To show all the labels?

`kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml`?

**Comment by sftim** [comment]
@enj would you be willing to propose those changes in a follow-up PR? I especially want the essentials to land, even as I also want this new page to see improvements.

**Comment by sftim** [comment]
We should get a tech review on this one.

**Comment by enj** [comment]
> We should get a tech review on this one.
> 

@deads2k and @lavalamp are good candidates here

File: content/en/docs/concepts/architecture/leases.md (hunk 2 above)
Comment (andrewsykim) [reply content/en/docs/concepts/architecture/leases.md:48]: done

File: content/en/docs/concepts/architecture/leases.md [hunk 3]
```diff
@@ -38,3 +38,40 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+```
+
+The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be
+configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname
+will take over existing Leases using a new holder identity, as opposed to instantiating new lease objects. You can check the
+hostname used by kube-apisever by checking the value of the `kubernetes.io/hostname` label:
+
+```
+$ kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml
+apiVersion: coordination.k8s.io/v1
```
Comment (enj) [inline content/en/docs/concepts/architecture/leases.md:60]: Can you split this into two code blocks, one with `shell` syntax, the other with `yaml` syntax?

**Review by enj — COMMENTED** [review]
One minor nit.

File: content/en/docs/concepts/architecture/leases.md (hunk 3 above)
Comment (andrewsykim) [reply content/en/docs/concepts/architecture/leases.md]: done!

**Comment by leonardpahlke** [comment]
/cc @deads2k @lavalamp
PTAL, thanks!

File: content/en/docs/concepts/architecture/leases.md [hunk 4]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
 discover how many instances of `kube-apiserver` are operating the Kubernetes control plane.
 Existence of kube-apiserver leases enables future capabilities that may require coordination between
 each kube-apiserver.
+
+You can inspect Leases owned by each kube-apiserver by checking for lease objects in the `kube-system` namespace
+with the name `kube-apiserver-<sha256-hash>`. Alternatively you can use the label selector `k8s.io/component=kube-apiserver`:
+
+```shell
+$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
+NAME                                        HOLDER                                                                           AGE
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:47]: Please split this - see [don't include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt) and [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

File: content/en/docs/concepts/architecture/leases.md [hunk 5]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
+kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
+kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
+```
+
+The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:53]: Which is true:
```suggestion
The SHA256 hash used in the Lease name is based on the OS hostname as seen by kube-apiserver. 
The API server coerces its hostname string to lowercase before hashing it. Each kube-apiserver
should be
```
or
```suggestion
The SHA256 hash used in the Lease name is based on the OS hostname as seen by kube-apiserver. 
Each kube-apiserver should be
```
?

File: content/en/docs/concepts/architecture/leases.md [hunk 6]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+```
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  creationTimestamp: "2022-11-30T15:37:15Z"
+  labels:
+    k8s.io/component: kube-apiserver
+    kubernetes.io/hostname: kind-control-plane
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:69]: https://kubernetes.io/docs/reference/labels-annotations-taints/ lists that this label is used on Nodes (only)

Either:
- fix the Kubernetes code not to label Leases with this official label
  - allowed, but a bit weird to make up a new one that is so similar
- update https://kubernetes.io/docs/reference/labels-annotations-taints/ to record that it is also used for leases that are associated with nodes

File: content/en/docs/concepts/architecture/leases.md [hunk 7]
```diff
@@ -29,7 +29,7 @@ This is used by control plane components like `kube-controller-manager` and `kub
 HA configurations, where only one instance of the component should be actively running while the other
 instances are on stand-by.
 
-## kube-apiserver identity
+## API Server Identity
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:32]:
```suggestion
## API server identity
```
or
```suggestion
## kube-apiserver identity {#api-server-identity}
```
(essentially: retain the original). Only kube-apiserver implements this mechanism as far as I know.

File: content/en/docs/concepts/architecture/leases.md [hunk 8]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+$ kubectl -n kube-system get lease -l k8s.io/component=kube-apiserver
+NAME                                        HOLDER                                                                           AGE
+kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a   kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4   5m33s
+kube-apiserver-dz2dqprdpsgnm756t5rnov7yka   kube-apiserver-dz2dqprdpsgnm756t5rnov7yka_84f2a85d-37c1-4b14-b6b9-603e62e4896f   4m23s
+kube-apiserver-fyloo45sdenffw2ugwaz3likua   kube-apiserver-fyloo45sdenffw2ugwaz3likua_c5ffa286-8a9a-45d4-91e7-61118ed58d2e   4m43s
+```
+
+The SHA256 hash used in the lease name is based on the OS hostname as seen by kube-apiserver. Each kube-apiserver should be
+configured to use a hostname that is unique within the cluster. New instances of kube-apiserver that use the same hostname
+will take over existing Leases using a new holder identity, as opposed to instantiating new lease objects. You can check the
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:55]:
```suggestion
will take over existing Leases using a new holder identity, as opposed to instantiating new Leases.
You can check the
```

File: content/en/docs/concepts/architecture/leases.md [hunk 9]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+hostname used by kube-apisever by checking the value of the `kubernetes.io/hostname` label:
+
+```shell
+$ kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:59]: (nit)
```suggestion
kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml
```
Maybe use a JSON path to get just the one label?

File: content/en/docs/concepts/architecture/leases.md [hunk 10]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+    k8s.io/component: kube-apiserver
+    kubernetes.io/hostname: kind-control-plane
+  name: kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a
+  namespace: kube-system
+  resourceVersion: "18171"
+  uid: d6c68901-4ec5-4385-b1ef-2d783738da6c
+spec:
+  holderIdentity: kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4
+  leaseDurationSeconds: 3600
+  renewTime: "2022-11-30T18:04:27.912073Z"
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:77]: (nit) Ideally, pick a date that is close to the v1.26 release.

File: content/en/docs/concepts/architecture/leases.md [hunk 11]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
 discover how many instances of `kube-apiserver` are operating the Kubernetes control plane.
 Existence of kube-apiserver leases enables future capabilities that may require coordination between
 each kube-apiserver.
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:40]:
```suggestion
The API server creates these leases in the `kube-system` namespace.
The existence of kube-apiserver Leases also enables adding future capabilities to Kubernetes,
that may require coordination between each kube-apiserver (for example, during cluster
upgrades).
```

File: content/en/docs/concepts/architecture/leases.md [hunk 12]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+
+You can inspect Leases owned by each kube-apiserver by checking for lease objects in the `kube-system` namespace
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:42]:
```suggestion
You can inspect Leases owned by each kube-apiserver by checking for Leases in the `kube-system` namespace
```

File: content/en/docs/concepts/architecture/leases.md [hunk 13]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+
+```shell
+$ kubectl -n kube-system get lease kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a -o yaml
+```
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  creationTimestamp: "2022-11-30T15:37:15Z"
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:66]: (nit) Ideally, pick a date that is close to the v1.26 release.

File: content/en/docs/concepts/architecture/leases.md [hunk 14]
```diff
@@ -38,3 +38,43 @@ rest of the system. While not particularly useful on its own, this provides a me
...
+  namespace: kube-system
+  resourceVersion: "18171"
+  uid: d6c68901-4ec5-4385-b1ef-2d783738da6c
+spec:
+  holderIdentity: kube-apiserver-c4vwjftbvpc5os2vvzle4qg27a_9cbf54e5-1136-44bd-8f9a-1dcd15c346b4
+  leaseDurationSeconds: 3600
+  renewTime: "2022-11-30T18:04:27.912073Z"
+```
+
+Expired leases from kube-apiservers that no longer exist are garbage collected by new kube-apiservers after 1 hour.
```
Comment (sftim) [inline content/en/docs/concepts/architecture/leases.md:80]: Is this right:
```suggestion
### API server Lease garbage collection

Expired Leases from kube-apiservers that no longer exist are garbage collected by live kube-apiservers.
The control plane leaves the expired Leases for at least one hour before removing them.
```
?

We should update https://kubernetes.io/docs/concepts/architecture/garbage-collection/ to link to that heading, which is why I added it.

**Review by sftim — COMMENTED** [review]
Some more feedback (see inline comments)

---

The label `k8s.io/component` is not registered. Please document it in https://kubernetes.io/docs/reference/labels-annotations-taints/ 

(if feasible: switch to using `control-plane.kubernetes.io/component`, **before the v1.26 release**, so that end users don't confuse it with the [`app.kubernetes.io/component`](https://kubernetes.io/docs/reference/labels-annotations-taints/#app-kubernetes-io-component) label).

We can also release with this and then deprecate `k8s.io/component` in v1.27 but that's a bit of a shame to have to do.

---

Optionally, document what happens when an API server runs as a Pod (eg for nesting a control plane).

That might be best saved for a post-release blog for the release when this feature graduates to stable.

File: content/en/docs/concepts/architecture/leases.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/concepts/architecture/leases.md:48]: See https://github.com/kubernetes/website/pull/38196#discussion_r1037486056

File: content/en/docs/concepts/architecture/leases.md (hunk 1 above)
Comment (sftim) [reply content/en/docs/concepts/architecture/leases.md:54]: See https://github.com/kubernetes/website/pull/38196#discussion_r1037486056

File: content/en/docs/concepts/architecture/leases.md (hunk 7 above)
Comment (enj) [reply content/en/docs/concepts/architecture/leases.md:32]: All API servers with storage are expected to implement this (i.e. aggregated API servers and Kube API server).  This is currently broken outside of KAS though.

**Comment by sftim** [comment]
Good enough for alpha, I reckon.

/approve

File: content/en/docs/concepts/architecture/leases.md (hunk 7 above)
Comment (enj) [reply content/en/docs/concepts/architecture/leases.md:32]: > OK. Presumably the expectation only comes in once the feature is GA?

This was something we missed by accident.  It should have been part of the beta.

**Comment by enj** [comment]
> Good enough for alpha, I reckon.

Except the feature is beta 😂

**Comment by sftim** [comment]
More polish is welcome.

**Pull request merged**